### PR TITLE
Add extra RPC endpoints for operator

### PR DESCRIPTION
**Pull request by jules**
## Motivation

Adding extra RPC endpoints to the operator in order to facilitate more data collection from mining pools

Massive thanks to @HarukaMa for doing the preliminary work on this!

### CODE REVIEW

File: src/rpc/documentation/public_endpoints/getprovers.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# Get Provers
+Returns the addresses of all provers that are registered on an operator.
```
Comment (ljedrz) [inline src/rpc/documentation/public_endpoints/getprovers.md]:
```suggestion
Returns the Aleo addresses of all provers that are registered with an operator.
```

File: src/rpc/documentation/public_endpoints/getprovers.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+### Arguments
+
+None
+
+### Response
+
+| Parameter |  Type  |                Description                 |
+|:---------:|:------:|:------------------------------------------:|
+| `result`  | array  | All of the addresses known to the operator |
```
Comment (ljedrz) [inline src/rpc/documentation/public_endpoints/getprovers.md]: known or actively registered?

File: src/rpc/documentation/public_endpoints/getshares.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
+# Get Shares
+Returns the amount of shares submitted to an operator in total.
```
Comment (ljedrz) [inline src/rpc/documentation/public_endpoints/getshares.md]: nit
```suggestion
Returns the total number of shares submitted to an operator.
```

File: src/rpc/documentation/public_endpoints/getshares.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+### Arguments
+
+None
+
+### Response
+
+| Parameter |  Type  |                       Description                       |
+|:---------:|:------:|:-------------------------------------------------------:|
+| `result`  |  u64   | The amount of shares submitted to the operator in total |
```
Comment (ljedrz) [inline src/rpc/documentation/public_endpoints/getshares.md]:
```suggestion
| `result`  |  u64   |  The total number of shares submitted to the operator   |
```

File: src/rpc/documentation/public_endpoints/getsharesforprover.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
+# Get Shares For Prover
+Returns the amount of shares submitted by a prover, given their address.
```
Comment (ljedrz) [inline src/rpc/documentation/public_endpoints/getsharesforprover.md]: nit
```suggestion
Returns the number of shares submitted by a prover, given their address.
```

File: src/rpc/documentation/public_endpoints/getsharesforprover.md [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+| Parameter |  Type  | Required |          Description           |
+|:---------:|:------:|:--------:|:------------------------------:|
+| `prover`  | string |   Yes    | The aleo address of the prover |
+
+### Response
+
+| Parameter |  Type  |                 Description                  |
+|:---------:|:------:|:--------------------------------------------:|
+| `result`  |  u64   | The amount of shares submitted by the prover |
```
Comment (ljedrz) [inline src/rpc/documentation/public_endpoints/getsharesforprover.md]:
```suggestion
| `result`  |  u64   | The number of shares submitted by the prover |
```

File: src/rpc/rpc_impl.rs [hunk 7]
```diff
@@ -337,4 +341,26 @@ impl<N: Network, E: Environment> RpcFunctions<N> for RpcImpl<N, E> {
...
     }
+
+    /// Returns the amount of shares submitted by a given prover.
+    async fn get_shares_for_prover(&self, prover: Value) -> Result<u64, RpcError> {
+        let prover: Address<N> = serde_json::from_value(prover)?;
+        Ok(self.operator.get_shares_for_prover(&prover))
+    }
+
+    /// Returns the amount of shares submitted to an operator in total.
+    async fn get_shares(&self) -> Result<u64, RpcError> {
```
Comment (ljedrz) [inline src/rpc/rpc_impl.rs]:
```suggestion
    async fn get_shares(&self) -> u64 {
```
there is no room for error here, we should avoid it if possible

File: src/rpc/rpc_impl.rs [hunk 8]
```diff
@@ -337,4 +341,26 @@ impl<N: Network, E: Environment> RpcFunctions<N> for RpcImpl<N, E> {
...
+        let shares = self.operator.to_shares();
+        let mut res = 0;
+        for (_, share) in shares {
+            res += share.values().sum::<u64>();
+        }
+        Ok(res)
+    }
+
+    /// Returns a list of provers registered on an operator.
+    async fn get_provers(&self) -> Result<Value, RpcError> {
```
Comment (ljedrz) [inline src/rpc/rpc_impl.rs]:
```suggestion
    async fn get_provers(&self) -> Value {
```
same reasoning as with `get_shares`

File: src/rpc/rpc_impl.rs [hunk 9]
```diff
@@ -337,4 +341,26 @@ impl<N: Network, E: Environment> RpcFunctions<N> for RpcImpl<N, E> {
...
+        Ok(self.operator.get_shares_for_prover(&prover))
+    }
+
+    /// Returns the amount of shares submitted to an operator in total.
+    async fn get_shares(&self) -> Result<u64, RpcError> {
+        let shares = self.operator.to_shares();
+        let mut res = 0;
+        for (_, share) in shares {
+            res += share.values().sum::<u64>();
+        }
```
Comment (ljedrz) [inline src/rpc/rpc_impl.rs]:
```suggestion
        res = shares.iter().map(|(_, share)| share.values().sum()).sum();
```
(probably)

**Review by ljedrz — APPROVED** [review]
LGTM with a few suggestions

File: src/network/operator.rs [hunk 10]
```diff
@@ -226,6 +226,13 @@ impl<N: Network, E: Environment> Operator<N, E> {
         self.state.get_shares_for_prover(&self.ledger_reader, prover)
     }
 
+    ///
+    /// Returns all registered provers on this operator.
```
Comment (niklaslong) [inline src/network/operator.rs]:
```suggestion
    /// Returns all provers registered with this operator.
```

File: storage/src/state/operator.rs [hunk 11]
```diff
@@ -76,6 +81,11 @@ impl<N: Network> OperatorState<N> {
     pub fn remove_shares(&self, block_height: u32, coinbase_record: Record<N>) -> Result<()> {
         self.shares.remove_shares(block_height, coinbase_record)
     }
+
+    /// Returns a list of provers registered on an operator.
```
Comment (niklaslong) [inline storage/src/state/operator.rs]:
```suggestion
    /// Returns a list of provers registered with an operator.
```

**Review by niklaslong — APPROVED** [review]
Left some suggestions but otherwise LGTM 👍

**Review by raychu86 — APPROVED** [review]

**Review by howardwu — APPROVED** [review]

**Pull request merged**